pyvisa-py/highlevel.py

```diff
@@ -431,9 +470,146 @@ def unlock(self, session):
+        :param event_type: Logical event identifier.
+        :param handler: Interpreted as a valid reference to a handler to be installed by a client application.
+        :param user_handle: A value specified by an application that can be used for identifying handlers
+                            uniquely for an event type.
+        :returns: a handler descriptor which consists of three elements:
+                 - handler (a python callable)
+                 - user handle (a ctypes object)
+                 - ctypes handler (ctypes object wrapping handler)
+                 and return value of the library call.
+        :rtype: int, :class:`pyvisa.constants.StatusCode`
+        """
+        try:
+            sess = self.sessions[session]
+        except KeyError:
+            return StatusCode.error_invalid_object
+
+        return sess.install_handler(event_type, handler, user_handle)
+
+    def uninstall_handler(self, session, event_type, handler, user_handle=None):
+        """Uninstalls handlers for events.
+
+        Corresponds to viUninstallHandler function of the VISA library.
+
+        :param session: Unique logical identifier to a session.
+        :param event_type: Logical event identifier.
+        :param handler: Interpreted as a valid reference to a handler to be uninstalled by a client application.
+        :param user_handle: A value specified by an application that can be used for identifying handlers
+                            uniquely in a session for an event.
+        :return: return value of the library call.
+        :rtype: :class:`pyvisa.constants.StatusCode`
+        """
+        try:
+            sess = self.sessions[session]
+        except KeyError:
+            return StatusCode.error_invalid_object
+
+        return sess.uninstall_handler(event_type, handler, user_handle)
+
+    def wait_on_event(self, session, in_event_type, timeout):
+        """Waits for an occurrence of the specified event for a given session.
+
+        Corresponds to viWaitOnEvent function of the VISA library.
+
+        :param session: Unique logical identifier to a session.
+        :param in_event_type: Logical identifier of the event(s) to wait for.
+        :param timeout: Absolute time period in time units that the resource shall wait for a specified event to
+                        occur before returning the time elapsed error. The time unit is in milliseconds.
+        :return: - Logical identifier of the event actually received
+                 - A handle specifying the unique occurrence of an event
+                 - return value of the library call.
+        :rtype: - eventtype
+                - event
+                - :class:`pyvisa.constants.StatusCode`
+        """
+        try:
+            sess = self.sessions[session]
+        except KeyError:
+            return None, None, StatusCode.error_invalid_object
+
+        # FIXME: for now calling Session's own wait_on_event() is the only way
```

Comment:
I am not sure I follow why this is an issue. Could you elaborate ?

Reply:
This particular comment probably does not belong in there anymore, but the gist of it is I am unclear how to nicely implement the VI_HNDLR mechanism. Event handlers are sitting stored in PyVisaLib.handlers. They are supposed to be called by Sessions (async nonetheless, and that is a whole other can of worms) which should pass event information to them. This was a note for me that the only clean way for sessions to pass events back to the library is to return them from wait_on_event.

Reply:
Once again I feel lack I don't have a good enough understanding of VISA events. And of how this would translate in Python (is the ctype handler called in a different thread). If we go further in the direction of event support we should investigate making handler easier to use. In particular I am unclear how to make both backends behave in a similar manner. 
Ideally I would like top be able to pass a python function as handler but that would require some work in the case of the ni backend.

Reply:
I feel implementing event handlers correctly in pyvisa-py is going to be very difficult for little gain. This part of VISA seems to be best implemented at driver level with callbacks into the app.

For example, I have no idea how to have the VI_HNDLR mechanism detect REN line changes of a GPIB board without continuously polling in a separate thread, and even with all the polling a quick pair of on-off events could easily be skipped completely.

I suggest testing and stabilizing the VI_QUEUE mechanism first. It is synchronous and simple. It only needs to implement Session.wait_on_event() which empties the session's own event queue or waits for an event to happen. That alone would already cover most of GPIBInstrument synchronization needs.

Reply:
I do agree let's focus on the queue mechanism that is easier. It is just that diving into the pyvisa side of handlers made me realize that they may be some work to there to to make it more usable.

Reply:
Let me know if I can help on the pyvisa front.

Reply:
Currently my main difficulty is to access the hardware required for testing. So if you can help with testing changes it would be great. I have some open PRs both on pyvisa-py and pyvisa that can use a round of testing. 
For the problem of handlers, I need to find time to look more in details how they currently work. Also since people do not generally complain about them, this is rather low priority.
Out of curiosity as one of the few recurrent contributor would you be interested in joining the organization ?

Reply:
I got the proper privileges so I will invite you directly @tivek.

Reply:
> @tivek ping me when this is ready for further reviews.

Right now I feel GPIB event queues are done. It still needs another round of testing and comparing with the NI backend. Right now I only have access to linux-gpib, but during the winter break I will most likely get my hands on a Windows machine with NI hardware.

Apart from that there is the event handler rework on the pyvisa side as discussed above: https://github.com/tivek/pyvisa/tree/expose_events. I should have a WIP pull request soon.

Reply:
> I got the proper privileges so I will invite you directly @tivek.

Great, thank you.